kinode/src/eth/provider.rs

```diff
@@ -1,189 +1,471 @@
+    Ok(())
 }
 
 /// Executed as a long-lived task. The JoinHandle is stored in the `connections` map.
 /// This task is responsible for connecting to the ETH RPC provider and streaming logs
 /// for a specific subscription made by a process.
 async fn handle_subscription_stream(
-    our: Arc<String>,
-    provider: Provider<Ws>,
-    filter: Filter,
+    our: String,
+    sub_id: u64,
+    mut rx: RawSubscription,
     target: Address,
+    rsvp: Option<Address>,
     send_to_loop: MessageSender,
 ) -> Result<(), EthError> {
-    let mut stream = match provider.subscribe_logs(&filter).await {
-        Ok(s) => s,
+    match rx.recv().await {
         Err(e) => {
-            return Err(EthError::ProviderError(e.to_string()));
+            return Err(EthError::SubscriptionClosed(sub_id))?;
         }
-    };
+        Ok(value) => {
+            // this should not return in case of one failed event?
+            let event: SubscriptionResult = serde_json::from_str(value.get()).map_err(|_| {
+                EthError::RpcError("eth: failed to deserialize subscription result".to_string())
+            })?;
+            send_to_loop
+                .send(KernelMessage {
+                    id: rand::random(),
+                    source: Address {
+                        node: our,
+                        process: ETH_PROCESS_ID.clone(),
+                    },
+                    target: target.clone(),
+                    rsvp: rsvp.clone(),
+                    message: Message::Request(Request {
+                        inherit: false,
+                        expects_response: None,
+                        body: serde_json::to_vec(&EthSub {
+                            id: sub_id,
+                            result: event,
+                        })
+                        .unwrap(),
+                        metadata: None,
+                        capabilities: vec![],
+                    }),
+                    lazy_load_blob: None,
+                })
+                .await
+                .unwrap();
+        }
+    }
+    Err(EthError::SubscriptionClosed(sub_id))
+}
 
-    while let Some(event) = stream.next().await {
-        send_to_loop
-            .send(KernelMessage {
-                id: rand::random(),
-                source: Address {
-                    node: our.to_string(),
-                    process: ETH_PROCESS_ID.clone(),
-                },
-                target: target.clone(),
-                rsvp: None,
-                message: Message::Request(Request {
-                    inherit: false,
-                    expects_response: None,
-                    body: serde_json::to_vec(&EthSubEvent::Log(event)).unwrap(),
-                    metadata: None,
-                    capabilities: vec![],
-                }),
-                lazy_load_blob: None,
-            })
-            .await
-            .unwrap();
+// todo, always send errors or no? general runtime question for other modules too.
```

Comment:
only send error if response is expected

Reply:
ok, I think I should make the case of a closed subscription special, as a process probably would like to know that even if subs have no expects_response

Reply:
well yes, that's a Request isn't it